src/uu/cp/src/cp.rs

```diff
@@ -1586,6 +1608,50 @@ fn copy_file(
         return Ok(());
     }
 
+    // issue 5031 case
+    // touch a && ln -s a b && mkdir c
+    // cp --preserve=links -R -H a b c
+    if let Some(base_name) = source.file_name() {
+        let base_source: PathBuf = base_name.into();
+        if let Some(new_source) = should_hard_linked_files.get(&base_source) {
+            std::fs::hard_link(new_source, dest)?;
+            return Ok(());
+        }
+    }
+    if source.is_symlink()
+        && options.dereference(source_in_command_line)
+        && options.preserve_hard_links()
```

Comment:
It would be nice to have some more comments in this section. Also maybe some of the names need some updates. For example, `preserve_hard_links` seems to not just be about hard links anymore.